### PR TITLE
Validate result collection keys

### LINKED ISSUE #699: restrict allowed characters in `Collection` keys
Because `Collection` keys are often used in filenames, we should restrict the allowed characters to avoid characters that are can be problematic in file paths. Our initial set of allowed characters (which may be expanded in the future if needed) will be we will allow `A-Z`, `a-z`, `0-9`, `+`, `-`, `_`, and `.`. 

If a value that will be used as a key contains any of these characters, the actions should fail with an error as early as possible. Given that we will ideally fail before work is done, it may make sense to have a public validate function that takes a list of keys and fails with an informative error message if any disallowed characters are present (e.g., a static `Collection.validate_keys` method). This could then be called when the keys are first defined. 

I initially implemented this as part of https://github.com/qiime2/q2-feature-table/pull/277 for an action that returns a `Collection` - see [this commit](https://github.com/gregcaporaso/q2-feature-table/commit/c23848820ea16faf270a941c6cb87e30b4ed9899).

### PR DESCRIPTION
**Pull request by Oddant1**
Ensure that result collection keys only contain the following characters: `A-Z`, `a-z`, `0-9`, `+`, `-`, `_`, and `.`.

Close #699

### CODE REVIEW

File: qiime2/sdk/util.py
Comment (gregcaporaso) [inline qiime2/sdk/util.py:128]:
```suggestion
                       ' characters: A-Z, a-z, 0-9, +, -, _, and .\n'
                       'The following key contains disallowed '
                       f'characters: {key}')
```

File: qiime2/sdk/tests/test_result.py
Comment (gregcaporaso) [inline qiime2/sdk/tests/test_result.py:594]:
```suggestion
                ' characters:.*valid key'):
```

File: qiime2/sdk/tests/test_result.py
Comment (gregcaporaso) [inline qiime2/sdk/tests/test_result.py:603]:
```suggestion
                ' characters:.*valid key'):
```

File: qiime2/sdk/util.py
Comment (gregcaporaso) [inline qiime2/sdk/util.py:125]: I recommend adding a couple of unit tests for this function itself, especially since it's public facing the in sdk.

**Review by gregcaporaso — CHANGES_REQUESTED** [review]
This looks great @Oddant1 - just a couple of quick suggestions.

**Review by gregcaporaso — APPROVED** [review]
@Oddant1, I ended up pushing a couple of commits while working on https://github.com/qiime2/q2-feature-table/pull/277. This all looks good to me, but you should review my changes.

**Review by lizgehret — COMMENTED** [review]
this lgtm @Oddant1 - once merge conflicts are resolved, i can get this one merged!

**Comment by gregcaporaso** [comment]
I'm not sure why that flake8 issue came up here - should I move that commit over to a separate pull request?

**Review by lizgehret — APPROVED** [review]

**Pull request merged**
